### ISSUE TITLE
Verify affinity between DS220+ environment and G1GC

### ISSUE THREAD

**Issue by tesshucom** [tesshucom/jpsonic#2342]
Prerequisites: #2310

### Overview

DS220+ has dual core environment (Intel Celeron J4025).

Smaller media servers the JVM will decide that Serial GC is appropriate for the Gecage Collector. In the environment of the scale envisaged by Jpsonic, we will investigate whether a better effect can be obtained by explicitly specifying G1.

### Details

In v112.2.0, the default Java options provided in <a href="https://raw.githubusercontent.com/jpsonic/jpsonic/master/install/docker/production.synology.yml">production.synology.yml</a> are:

``JAVA_OPTS: '-Xms512m -Xmx512m -XX:MaxGCPauseMillis=500'``

#### About MaxGCPauseMillis

 - MaxGCPauseMillis is a valid property for both serial and G1. Relaxing MaxGCPauseMillis will reduce unnecessary GC activity and increase scan throughput in Jpsonic.
 - Sensory responses other than during scanning are not affected. Fortunately, even if you compete with GC, the additional time most of the time is only a few tens of milliseconds. This is not a problem in most cases. Because, at least personal media servers are a type of service that does not require severe responsiveness. The only caveat is DLNA (not UPnP) with 500ms response rules.

<details>
<summary>Verification summary of scan time for v112.2</summary>

Scan throughput is improved by relaxing MaxGCPauseMillis.

>  - Dummy data used in #2310
>    - The playing time is 0 to reduce the influence of seek time.
>    - Contains impractical data to force kick some SQL
>      - An incorrect track number is registered , etc.
>    - It is possible to make relative comparisons between dummy data results. However, please note that the total scan time is not realistic.

100K | v112.1.0 (Xmx) | v112.1.8 (Xmx, Xms) | v112.1.8 (Xmx, Xms, MaxGCPauseMillis)
-- | -- | -- | --
Duration All - 1 | 0:04:45.125 | 0:03:33.484 | 0:02:53.684
Duration All - 2 | 0:04:45.069 | 0:03:31.001 | 0:02:49.776
Duration All - 3 | 0:04:49.920 | 0:03:30.754 | 0:02:51.883
Average | 0:04:46.705 | 0:03:31.746 | 0:02:51.781

200K | v112.1.0 (Xmx) | v112.1.8 (Xmx, Xms) | v112.1.8 (Xmx, Xms, MaxGCPauseMillis)
-- | -- | -- | --
Duration All - 1 | 0:16:41.385 | 0:06:51.759 | 0:06:19.619
Duration All - 2 | 0:16:30.339 | 0:07:22.592 | 0:06:17.991
Duration All - 3 | 0:16:30.348 | 0:07:21.110 | 0:06:17.552
Average | 0:16:34.024 | 0:07:11.820 | 0:06:18.387

300K | v112.1.0 (Xmx) | v112.1.8 (Xmx, Xms) | v112.1.8 (Xmx, Xms, MaxGCPauseMillis)
-- | -- | -- | --
Duration All - 1 | 0:21:11.014 | 0:10:45.815 | 0:09:44.301
Duration All - 2 | - | 0:10:32.806 | 0:09:45.811
Duration All - 3 | - | 0:10:29.824 | 0:09:42.887
Average | 0:21:11.014 | 0:10:36.148 | 0:09:44.333

400K | v112.1.0 (Xmx) | v112.1.8 (Xmx, Xms) | v112.1.8 (Xmx, Xms, MaxGCPauseMillis)
-- | -- | -- | --
Duration All - 1 | 0:29:44.527 | 0:14:07.848 | 0:13:21.062
Duration All - 2 | 0:30:04.902 | 0:14:19.541 | 0:13:17.026
Duration All - 3 | 0:29:47.017 | 0:14:12.537 | 0:13:19.262
Average | 0:29:52.149 | 0:14:13.309 | 0:13:19.117

</details>

Xms,Xmx,MaxGCPauseMillis are "suggested properties" for Jpsonic no matter what GC is chosen.

#### Default Garbage Collector on DS220+&Jpsonic

For DS220+, when the above option is used, Serial GC is selected as shown below.

<details>
<summary>gc log on DS220+</summary>

```
[info][gc] Using Serial
[info][gc,init] Version: 17.0.8+7 (release)
[info][gc,init] CPUs: 2 total, 2 available
[info][gc,init] Memory: 1024M
[info][gc,init] Large Page Support: Disabled
[info][gc,init] NUMA Support: Disabled
[info][gc,init] Compressed Oops: Enabled (32-bit)
[info][gc,init] Heap Min Capacity: 512M
[info][gc,init] Heap Initial Capacity: 512M
[info][gc,init] Heap Max Capacity: 512M
[info][gc,init] Pre-touch: Disabled
```

</details>

Current validation is based on this GC, so additional validation will be done eith G1GC.

### Goal

The goal is to get a verdict on whether it is good or bad to force G1GC on the DS220+(≒ Dual Core environment that is not so high spec).

 - Default Java Options provided in production.synology.yml should be updated if better results are observed.
 - If the result does not change much, it should be left as it is. Explicitly specifying GC does not necessarily lead to better results. Explicit GC selection can only be offered to users if you are confident in the results.
   - Delegating GC selection to the JVM provides more versatility. If there is a point that is slightly different from the tuning required in the business scene, it would be this. For OSS personal media servers, it is common practice to deploy the same product in various environments such as Desktop, VPS, NAS, Raspberry Pi, etc.

### Non-goal

 - The purpose is not to confirm which is better, Serial GC or G1 GC. This is an affinity issue between G1GC and a small-scale app that assumes DS220+.

### Procedure

 - Run the same validation as before, adding the condition that G1 GC is used. Add `-XX:+UseG1GC`. 
   -  e.g. -Xms512m -Xmx512m -XX:MaxGCPauseMillis=500 -XX:+UseG1GC
 - Serial GC is selected when using the 10K and 20K settings. Therefore, only 10K and 20K will be additionally verified.

___

### Result

<details>
<summary>Dummy data with playing time 0 (Used in #2310)</summary>

100K | 1 | 2 | 3
-- | -- | -- | --
Total scan time | 0:02:44.400 | 0:02:54.243 | 0:02:56.521
Durtion of PARSE_FILE_STRUCTURE | 0:01:54.500 | 0:02:03.441 | 0:02:06.789
Throughput (%) | 96.288 | 96.499 | 96.544
Latency - Avg Pause GC Time (ms) | 17.3 | 17.6 | 17.6
Latency - Max Pause GC Time (ms) | 70 | 60 | 80
Memory Leak | None. | None. | None.
Consecutive Full GC | None. | None. | None.
Long Pause | None. | None. | None.

200K | 1 | 2 | 3
-- | -- | -- | --
Total scan time | 0:06:50.598 | 0:06:46.477 | 0:06:47.988
Durtion of PARSE_FILE_STRUCTURE | 0:04:55.765 | 0:04:53.641 | 0:04:55.469
Throughput (%) | 97.483 | 97.445 | 97.484
Latency - Avg Pause GC Time (ms) | 32.7 | 32.3 | 32.7
Latency - Max Pause GC Time (ms) | 160 | 160 | 150
Memory Leak | None. | None. | None.
Consecutive Full GC | None. | None. | None.
Long Pause | None. | None. | None.

</details>
<details>
<summary>Dummy data with playing time 5min (Used in #2310)</summary>

> Note that dummy data is used (for realistic data it is almost always a little faster)

5min*100K/G1| 1
-- | --
Total scan time | 0:31:40.416
Durtion of PARSE_FILE_STRUCTURE | 0:30:56.429
Throughput (%) | 99.621
Latency - Avg Pause GC Time (ms) | 16.8
Latency - Max Pause GC Time (ms) | 80
Memory Leak | None.
Consecutive Full GC | None.
Long Pause | None.

![image](https://github.com/tesshucom/jpsonic/assets/27724847/93afccd9-750c-4d83-ac61-60becaa02eb2)

5min*100K/Serial | 1
-- | --
Total scan time | 0:20:58.742
Durtion of PARSE_FILE_STRUCTURE | 0:20:04.845
Throughput (%) | 98.321
Latency - Avg Pause GC Time (ms) | 56.1
Latency - Max Pause GC Time (ms) | 1 sec 248 ms
Memory Leak | None.
Consecutive Full GC | None.
Long Pause | None.

![image](https://github.com/tesshucom/jpsonic/assets/27724847/7d5763f9-11c1-4fb2-82f9-eb793a437631)

</details>

### Summary

 - We should not specify Garbage Collector from Java Option. Delegating to automatic JVM selection will give better results overall.
 - Therefore, the Java options listed in production.synology.yml provided with v112.2.0 will not change.
   - For the DS220+, 100K and 200K will select Serial, and 300K and above memory settings will select G1. It is better to follow it.
   - For your VPS, Desktop machines, or Raspberry Pi, G1 may always be the choice. Of course, in that environment it is better to use it as is.

Forcing G1 GC when Serial GC is selected in a Dual Core environment would significantly increase the processing time of the PARSE_FILE_STRUCTURE step due to overhead. On the other hand, the processing efficiency is slightly improved in the SQL-based process in the latter half of the scan process. In a Full SSD environment, this disadvantage is offset, so even if the JVM determines that Serial GC is appropriate, forcing G1 GC may yield good results. However, if you add in more than 100K Flac songs and Synology's OS area, you will need a 4TB class SSD, easily exceeding the price of the DS220+. By common sense, that is not a realistic environment assumption. When it comes to Jpsonic, there is no problem in saying that the optimal solution is to delegate GC selection to the JVM.

<details>
<summary>By the way, some thoughts on parallelization</summary>

There is also an idea that the PARSE_FILE_STRUCTURE process time(seak time of HDD) can be reduced somewhat by parallelizing the process. However, it is **short-sighted**.

 - At least the verification considering CPU, memory and HDD performance should be done using G1 and Serial GC. Even Synology's Plus series uses Dual Core. When it comes to popular NAS machines, Serial GC is more likely to be used. At least for the last few years. Alternatively, this trend will continue until the global semiconductor shortage is resolved and NAS devices that use Quad Core become popular all over the world.
 - Resource control can easily collapse without proper parallelization in processing blocks and proper stack management. Depending on the environment, it may work for some people. For some, it's a nightmare.
 - This cannot be resolved through discussion, but requires proper design and verification. At least considering the current lineup of major NAS products, it should be a single-threaded scan. Multithreaded scanning should be an Option that the user optionally enables. Even if multi-threaded scanning is provided with sufficient design, implementation, and verification, users will dare to choose single-threaded scanning when compatibility with other services such as NAS needs to be considered. Users who understand the nearly zero-sum relationship between speed and resource usage may do so.

A solution without verification seems like a difficult topic to establish. 

</details>

**Comment by tesshucom** [comment]
There was a problem with the measurement procedure, so it will be re-measured.

**Comment by tesshucom** [comment]
Solved by #2345.
